nats-compiler-plugin/src/main/java/io/ballerina/stdlib/nats/plugin/NatsFunctionValidator.java

```diff
@@ -0,0 +1,223 @@
+            validateErrorParam(functionDefinitionNode, parameters.get(1));
+        }
+        if (parameters.size() < 2) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.MUST_HAVE_MESSAGE_AND_ERROR,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+        if (parameters.size() > 2) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.ONLY_PARAMS_ALLOWED_ON_ERROR,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+    }
+
+    private void validateFirstParam(FunctionDefinitionNode functionDefinitionNode,
+                                    ParameterNode parameterNode) {
+        RequiredParameterNode requiredParameterNode = (RequiredParameterNode) parameterNode;
+        Node parameterTypeNode = requiredParameterNode.typeName();
+        if (!parameterTypeNode.kind().equals(SyntaxKind.QUALIFIED_NAME_REFERENCE)) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(
+                    CompilationErrors.INVALID_FUNCTION_PARAM_MESSAGE,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.location()));
+        } else {
+            Token modulePrefix = ((QualifiedNameReferenceNode) parameterTypeNode).modulePrefix();
+            IdentifierToken identifierToken = ((QualifiedNameReferenceNode) parameterTypeNode).identifier();
+            if (!modulePrefix.text().equals(PluginConstants.PACKAGE_PREFIX) ||
+                    !identifierToken.text().equals(PluginConstants.MESSAGE)) {
+                context.reportDiagnostic(PluginUtils.getDiagnostic(
+                        CompilationErrors.INVALID_FUNCTION_PARAM_MESSAGE,
+                        DiagnosticSeverity.ERROR, functionDefinitionNode.location()));
+            }
+        }
+    }
+
+    private void validateReturnTypeErrorOrNil(FunctionDefinitionNode functionDefinitionNode) {
+        Optional<ReturnTypeDescriptorNode> returnTypes = functionDefinitionNode.functionSignature().returnTypeDesc();
+        if (returnTypes.isPresent()) {
+            ReturnTypeDescriptorNode returnTypeDescriptorNode = returnTypes.get();
+            Node returnNodeType = returnTypeDescriptorNode.type();
+            String returnType = returnNodeType.toString().split(" ")[0];
+            if (!returnType.equals(PluginConstants.ERROR_OR_NIL) &&
+                    !returnType.equals(PluginConstants.NIL_OR_ERROR) &&
+                    !returnType.equals(PluginConstants.NATS_ERROR_OR_NIL) &&
+                    !returnType.equals(PluginConstants.NIL_OR_NATS_ERROR)) {
+                context.reportDiagnostic(PluginUtils.getDiagnostic(
+                        CompilationErrors.INVALID_RETURN_TYPE_ERROR_OR_NIL,
+                        DiagnosticSeverity.ERROR, functionDefinitionNode.location()));
+            }
```

Comment:
For this analysis, it'd be easier and more consistent to use the semantic API. 
- Lookup the symbol of the function using `symbol()`
- Get the return type from that symbol
- Check if it's a union type, if so, validate the member types of that union